torch/csrc/jit/codegen/cuda/ir_cloner.h

```diff
@@ -0,0 +1,91 @@
+
+#pragma once
+
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <torch/csrc/jit/codegen/cuda/dispatch.h>
+
+#include <unordered_map>
+#include <vector>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+
+class Fusion;
+
+// Clones nodes from an exiting Fusion
+class TORCH_CUDA_API IrCloner : private OptInConstDispatch {
+  friend struct Statement;
+
+ public:
+  explicit IrCloner(Fusion* new_fusion) : fusion_(new_fusion) {}
+
+  Statement* clone(const Statement* statement);
+
+  template <class T>
+  T* clone(const T* node) {
+    return node ? clone(node->template as<Statement>())->template as<T>()
+                : nullptr;
+  }
+
+  template <class T>
+  std::vector<T*> clone(const std::vector<T*>& container) {
+    std::vector<T*> copy;
+    for (auto p : container) {
+      copy.push_back(clone(p));
+    }
+    return copy;
+  }
+
+  Fusion* fusion() const {
+    return fusion_;
+  }
+
+ private:
+  void registerClone(const Statement* src, Statement* clone);
+
+  void handle(const Statement*) override;
+  void handle(const Val*) override;
+  void handle(const Expr*) override;
+
+  void handle(const TensorDomain*) override;
+  void handle(const TensorView*) override;
+  void handle(const IterDomain*) override;
+  void handle(const TensorIndex*) override;
+
+  void handle(const Bool*) override;
+  void handle(const Float*) override;
+  void handle(const Half*) override;
+  void handle(const Int*) override;
+  void handle(const NamedScalar*) override;
+
+  void handle(const UnaryOp*) override;
+  void handle(const BinaryOp*) override;
+  void handle(const TernaryOp*) override;
+  void handle(const BroadcastOp*) override;
+  void handle(const ReductionOp*) override;
+
+  void handle(const ForLoop*) override;
+  void handle(const IfThenElse*) override;
+  void handle(const Allocate*) override;
+
+  void handle(const Split*) override;
+  void handle(const Merge*) override;
+
+ private:
+  // The destination Fusion container
+  Fusion* fusion_ = nullptr;
+
+  // The dispatch interface doesn't allow returning values from
```

Comment:
Could you use the mutate dispatch here?

Reply:
Good point. We'd need a new OptInConstMutator, plus change the mutations map to const* Statement -> Statement*

It could be done, but it may not be a net simplification for a single use. What do you think?

PS. Since we are on topic, why is OptInMutator::mutations mapping Val* and not Statement* ?

Reply:
I think we can definitely hold off on this for now and keep it in mind. Not sure on the P.S. as I don't use mutate too often, I think it's used in lowering though we could check if it makes any sense there, otherwise maybe just a mistake.